app/src/main/java/org/mozilla/focus/session/ui/SessionsAdapter.kt

```diff
@@ -65,8 +64,24 @@ class SessionsAdapter/* package */ internal constructor(private val fragment: Se
         return sessions!!.size + 1
     }
 
-    override fun onChanged(sessions: List<Session>?) {
-        this.sessions = ArrayList(sessions!!)
+    override fun onSessionAdded(session: Session) {
+        onUpdate(fragment.requireComponents.sessionManager.sessions)
+    }
+
+    override fun onSessionRemoved(session: Session) {
+        onUpdate(fragment.requireComponents.sessionManager.sessions)
+    }
+
+    override fun onSessionSelected(session: Session) {
+        onUpdate(fragment.requireComponents.sessionManager.sessions)
+    }
+
+    override fun onAllSessionsRemoved() {
+        onUpdate(fragment.requireComponents.sessionManager.sessions)
+    }
+
+    private fun onUpdate(sessions: List<Session>) {
+        this.sessions = sessions
```

Comment:
In general, it would be better to have DiffUtil calculate the diff and update the adapter when possible. There are non-trivial performance and animation differences with notifyDataSetChanged().